[UnifiedPDF] Context menu events should update selections

aprotyas opened the pull request:
#### 8099f94806f156869d7f806084efb542ea1658c9
<pre>
[UnifiedPDF] Context menu events should update selections
<a href="https://bugs.webkit.org/show_bug.cgi?id=268951">https://bugs.webkit.org/show_bug.cgi?id=268951</a>
<a href="https://rdar.apple.com/122503367">rdar://122503367</a>

Reviewed by Tim Horton and Simon Fraser.

This patch introduces support for selection updates based on context
menu events (right click or ctrl-click).

It does so by adding some logic for context menu events that shortcuts
the rest of our selection tracking operation. Specifically, we call into
updateCurrentSelectionForContextMenuEventIfNeeded, which generates a text
selection for the word under which the click is produced, unless of
course the click location already intersects with an existing selection.

Furthermore, we add a lastHandledEventWasContextMenuEvent flag to our
selection tracking data, which we consult when determining whether
selections should be painted with an active or inactive appearance. This
is relevant because interacting with context menus causes selections to
go inactive in shipping macOS.

Finally, we make some drive-by fixes where we opt to pass enumeration
values such as WebEventType by value and not as const references, since
they are POD types.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::windowActivityDidChange):
(WebKit::UnifiedPDFPlugin::paintPDFContent):
(WebKit::isContextMenuEvent):
(WebKit::UnifiedPDFPlugin::handleMouseEvent):
(WebKit::UnifiedPDFPlugin::handleContextMenuEvent):
(WebKit::UnifiedPDFPlugin::extendCurrentSelectionIfNeeded):
(WebKit::UnifiedPDFPlugin::beginTrackingSelection):
(WebKit::UnifiedPDFPlugin::updateCurrentSelectionForContextMenuEventIfNeeded):
(WebKit::UnifiedPDFPlugin::repaintOnSelectionActiveStateChangeIfNeeded):
(WebKit::UnifiedPDFPlugin::isSelectionActiveAfterContextMenuInteraction const):
(WebKit::UnifiedPDFPlugin::setCurrentSelection):
(WebKit::AnnotationTrackingState::startAnnotationTracking):
(WebKit::AnnotationTrackingState::finishAnnotationTracking):

Canonical link: <a href="https://commits.webkit.org/274347@main">https://commits.webkit.org/274347@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/969e56be3a56363a7cd3acee00649dbe12ef1814

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/38810 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/17741 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41155 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/41341 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/34463 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/20561 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15090 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/32537 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/39383 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/14929 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/33671 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/12966 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/12933 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/42617 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35226 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/34932 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/38762 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/13620 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11238 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/36977 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/15232 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8689 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/14898 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/14713 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1383,26 +1404,42 @@ static AffineTransform documentSpaceToPageSpaceTransform(const IntDegrees& pageR
+{
+    auto page = m_documentLayout.pageAtIndex(m_selectionTrackingData.startPageIndex);
+    if (!m_currentSelection || !(IntRect([m_currentSelection boundsForPage:page.get()]).contains(m_selectionTrackingData.startPagePoint)))
+        setCurrentSelection([page selectionForWordAtPoint:m_selectionTrackingData.startPagePoint]);
```
Is word granularity right, or does this need to use Lookup?

aprotyas replied:
Word granularity is wrong, but also not a relevant concept here, so I'll remove any changes to granularity tracking.

aprotyas replied:
Dropped in https://github.com/WebKit/WebKit/commit/969e56be3a56363a7cd3acee00649dbe12ef1814

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1067,6 +1068,13 @@ static AffineTransform documentSpaceToPageSpaceTransform(const IntDegrees& pageR
+
+static bool isContextMenuEvent(WebMouseEventButton button, OptionSet<WebEventModifier> modifiers)
+{
+    return button == WebMouseEventButton::Right || (button == WebMouseEventButton::Left && modifiers.contains(WebEventModifier::ControlKey));
```
Do we really have to code this here, instead of getting it from NSEvent somehow? I don't know if "controlKey" is true for all keyboards/user configs

aprotyas replied:
Ah good point, we should probably use something like `isContextClick` from `WebFrame`

https://github.com/WebKit/WebKit/blob/bec01ed582b08a515921746208c15f0a641f6de7/Source/WebKit/WebProcess/WebPage/WebFrame.cpp#L1207-L1215

aprotyas replied:
Fixed in https://github.com/WebKit/WebKit/commit/969e56be3a56363a7cd3acee00649dbe12ef1814

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1201,6 +1218,10 @@ static AffineTransform documentSpaceToPageSpaceTransform(const IntDegrees& pageR
+        m_selectionTrackingData.lastHandledEventWasContextMenuEvent = true;
+        // We make a repaint request if we've finished a context menu interaction.
+        // This allows us to switch to an inactive appearance for the selections.
+        m_contentsLayer->setNeedsDisplay();
```
Your comment could turn into self-documenting code if you wrapped up these two lines in a function, and had a corresponding function for the place where you had to comment:
```
        // Selections should be presented as inactive if we just interacted with a context menu.
        bool isVisibleAndActive = !m_selectionTrackingData.lastHandledEventWasContextMenuEvent;
```

So you could have:
`repaintOnSelectionActiveStateChange()` and `selectionActiveForContextMenu()` or something.

aprotyas replied:
Thanks for the suggestion. I'll try it. I knew there was a better way to present this than the pair of comments I dropped.

aprotyas replied:
Implemented your suggestion in [969e56b](https://github.com/WebKit/WebKit/commit/969e56be3a56363a7cd3acee00649dbe12ef1814) through:

```cpp
enum class ActiveStateChangeReason : uint8_t {
    HandledContextMenuEvent,
    BeganTrackingSelection,
    SetCurrentSelection,
    WindowActivityChanged,
};

void repaintOnSelectionActiveStateChangeIfNeeded(ActiveStateChangeReason);
bool isSelectionActiveAfterContextMenuInteraction() const;
```

The selection-related paint request logic is much cleaner now!

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -1383,26 +1404,42 @@ static AffineTransform documentSpaceToPageSpaceTransform(const IntDegrees& pageR
+    m_selectionTrackingData.selectionToExtendWith = nullptr;
+    // We make a repaint request if we've started tracking selection again after a context menu interaction.
+    // This allows us to toggle back to an active appearance for the selections.
+    if (std::exchange(m_selectionTrackingData.lastHandledEventWasContextMenuEvent, false))
```
Oh look here it is again (the selection state/repaint stuff)

aprotyas replied:
https://github.com/WebKit/WebKit/pull/24115#discussion_r1483851813

aprotyas committed:
[UnifiedPDF] Context menu events should update selections
https://bugs.webkit.org/show_bug.cgi?id=268951
rdar://122503367

Reviewed by Tim Horton and Simon Fraser.

This patch introduces support for selection updates based on context
menu events (right click or ctrl-click).

It does so by adding some logic for context menu events that shortcuts
the rest of our selection tracking operation. Specifically, we call into
updateCurrentSelectionForContextMenuEventIfNeeded, which generates a text
selection for the word under which the click is produced, unless of
course the click location already intersects with an existing selection.

Furthermore, we add a lastHandledEventWasContextMenuEvent flag to our
selection tracking data, which we consult when determining whether
selections should be painted with an active or inactive appearance. This
is relevant because interacting with context menus causes selections to
go inactive in shipping macOS.

Finally, we make some drive-by fixes where we opt to pass enumeration
values such as WebEventType by value and not as const references, since
they are POD types.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::windowActivityDidChange):
(WebKit::UnifiedPDFPlugin::paintPDFContent):
(WebKit::isContextMenuEvent):
(WebKit::UnifiedPDFPlugin::handleMouseEvent):
(WebKit::UnifiedPDFPlugin::handleContextMenuEvent):
(WebKit::UnifiedPDFPlugin::extendCurrentSelectionIfNeeded):
(WebKit::UnifiedPDFPlugin::beginTrackingSelection):
(WebKit::UnifiedPDFPlugin::updateCurrentSelectionForContextMenuEventIfNeeded):
(WebKit::UnifiedPDFPlugin::repaintOnSelectionActiveStateChangeIfNeeded):
(WebKit::UnifiedPDFPlugin::isSelectionActiveAfterContextMenuInteraction const):
(WebKit::UnifiedPDFPlugin::setCurrentSelection):
(WebKit::AnnotationTrackingState::startAnnotationTracking):
(WebKit::AnnotationTrackingState::finishAnnotationTracking):

Canonical link: https://commits.webkit.org/274347@main
